apps/server/src/modules/files-storage/client/s3-client.adapter.ts

```diff
@@ -189,37 +187,43 @@ export class S3ClientAdapter implements IStorageClient {
 		return this.client.send(req);
 	}
 
-	public async list(prefix: string, maxKeys?: number) {
-		this.logger.log({ action: 'list', params: { prefix, bucket: this.config.bucket } });
-
+	public async list(params: IListFiles) {
 		try {
-			let files: string[] = [];
-			let ret: ListObjectsV2CommandOutput | undefined;
+			this.logger.log({ action: 'list', params });
 
-			do {
-				const req = new ListObjectsV2Command({
-					Bucket: this.config.bucket,
-					Prefix: prefix,
-					ContinuationToken: ret?.NextContinuationToken,
-					MaxKeys: maxKeys && maxKeys - files.length,
-				});
+			const result = await this.listObjectKeysRecursive(params);
+
+			return result;
+		} catch (err) {
+			throw new InternalServerErrorException('S3ClientAdapter:listDirectory');
+		}
+	}
 
-				// Iterations are dependent on each other
-				// eslint-disable-next-line no-await-in-loop
-				ret = await this.client.send(req);
+	private async listObjectKeysRecursive(params: IListFiles) {
+		const { path, maxKeys, nextMarker } = params;
+		const files: string[] = params.files ? params.files : [];
 
-				const returnedFiles =
-					ret?.Contents?.filter((o) => o.Key)
-						.map((o) => o.Key as string) // Can not be undefined because of filter above
-						.map((key) => key.substring(prefix.length)) ?? [];
+		const req = new ListObjectsV2Command({
+			Bucket: this.config.bucket,
+			Prefix: path,
+			ContinuationToken: nextMarker,
+			MaxKeys: maxKeys,
+		});
 
-				files = files.concat(returnedFiles);
-			} while (ret?.IsTruncated && (!maxKeys || files.length < maxKeys));
+		const data = await this.client.send(req);
 
-			return files;
-		} catch (err) {
-			throw new InternalServerErrorException(err, 'S3ClientAdapter:list');
+		const returnedFiles =
+			data?.Contents?.filter((o) => o.Key)
```

Comment:
@marode-cap after speaking with @SevenWaysDP this part is h5p related. You have one of the comments to but it to a private method. But this method should be placed in the h5p. Maybe not as a private method it look more then a mapper. 

For me is the place where it is done not interessting. We can merge the code und you change it on thr-18, or if you like you can also take over the PR and do it on this place.

Reply:
Only the last part of that chain could be considered part of H5P (Removing the prefix from the filenames).
Filtering out files with no Key (I honestly don't know why S3 would even return those, but thats what the types say) should probably be a part of this function

Reply:
We have a ticket to move the s3-client in this we check what we want to return.